crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -5889,11 +5889,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_binary_expression_type(binary.into(), false, left_ty, right_ty, *op)
             .unwrap_or_else(|| {
                 if let Some(builder) = self.context.report_lint(&UNSUPPORTED_OPERATOR, binary) {
-                    builder.into_diagnostic(format_args!(
+                    let mut diag=builder.into_diagnostic(format_args!(
                         "Operator `{op}` is unsupported between objects of type `{}` and `{}`",
                         left_ty.display(self.db()),
                         right_ty.display(self.db())
                     ));
+
+                    if op == &ast::Operator::BitOr && left_ty.into_class_literal().is_some()
+                        && right_ty.into_class_literal().is_some() && Program::get(self.db()).python_version(self.db())
+                            < PythonVersion::PY310
+                    {
+                        diag.info("Note that `X | Y` PEP 604 union syntax is only available in Python 3.10 and later");
+                    }
```

Comment:
One could certainly construct cases where we emit this info even if the user is trying to do something that does not involve type hints, but it seems unlikely? And not too distracting even if?

On the contrary, there are probably also cases where we fail to emit this info due to the class-literal filter above. We could think about relaxing this (e.g. checking if `type` is in the MRO of both `left_ty` and `right_ty`, or further trying to detect other types that might appear in those positions), but I'm not sure if it's worth it?